File: geekbang/algorithm-models/6.2-并查集代码模板/DisjointSet.js
// JavaScript


class DisjointSet {
    constructor(n) {
        this.fa = [];
        for (let i = 0; i < n; i++) this.fa[i] = i;
    }


    find(x) {
        if (x == this.fa[x]) return x;
        return this.fa[x] = this.find(this.fa[x]);
    }


    unionSet(x, y) {
        x = this.find(x);
        y = this.find(y);
        if (x != y) this.fa[x] = y;
    }
};

File: geekbang/algorithm-models/6.2-并查集代码模板/DisjointSet.py
# Python


class DisjointSet:
    def __init__(self, n):
        self.fa = [i for i in range(n)]

    def find(self, x):
        if x == self.fa[x]:
            return x
        self.fa[x] = self.find(self.fa[x])
        return self.fa[x]

    def unionSet(self, x, y):
        x = self.find(x)
        y = self.find(y)
        if x != y:
            self.fa[x] = y
